modvendor/modvendor_test.go

```diff
@@ -15,18 +15,823 @@
+	// <relative path of modsrc2>, ...].
+	// Then the template can reference the third element of the list:
+	//   source = {{index 2}}
+	// If this looks like a hack to you, it's because it is.
+	wantFiles map[vendorPathSpec]fmt.Stringer
+}
+
+type hclconfig struct {
+	repo string
+	path string
+	data fmt.Stringer
+}
+
+type wantIgnoredVendor struct {
+	RawSource     string
+	ReasonPattern string
+}
+
+type wantReport struct {
+	Vendored []string
+	Ignored  []wantIgnoredVendor
+	Error    error
+}
+
+func TestModVendorAllRecursive(t *testing.T) {
+	tcases := []testcase{
+		{
+			name: "module with no remote deps",
+			layout: []string{
+				"g:module-test",
+			},
+			source: "git::file://{{.}}/module-test?ref=main",
+			configs: []hclconfig{
+				{
+					repo: "module-test",
+					path: "module-test/main.tf",
+					data: Module(
+						Labels("test"),
+						Str("source", "./modules/non-existent"),
+					),
+				},
+			},
+			wantVendored: []string{
+				"git::file://{{.}}/module-test?ref=main",
+			},
+		},
+		{
+			name: "module with HCL errors",
+			layout: []string{
+				"g:module-test",
+			},
+			source: "git::file://{{.}}/module-test?ref=main",
+			configs: []hclconfig{
+				{
+					repo: "module-test",
+					path: "module-test/main.tf",
+					data: bytes.NewBufferString("this is not a valid HCL file"),
+				},
+			},
+			wantVendored: []string{
+				"git::file://{{.}}/module-test?ref=main",
+			},
+			wantError: errors.E(hcl.ErrHCLSyntax),
+		},
+		{
+			name: "module with ignored remote deps",
+			layout: []string{
+				"g:module-test",
+			},
+			source: "git::file://{{.}}/module-test?ref=main",
+			configs: []hclconfig{
+				{
+					repo: "module-test",
+					path: "module-test/main.tf",
+					data: Module(
+						Labels("test"),
+						Str("source", "https://example.com/my-module"),
+					),
+				},
+			},
+			wantVendored: []string{
+				"git::file://{{.}}/module-test?ref=main",
+			},
+			wantIgnored: []wantIgnoredVendor{
+				{
+					RawSource:     "https://example.com/my-module",
+					ReasonPattern: "unsupported module source",
+				},
+			},
+		},
+		{
+			name:   "module not found",
+			source: "git::file://{{.}}/module-that-does-not-exists?ref=main",
+			wantIgnored: []wantIgnoredVendor{
+				{
+					RawSource:     "git::file://{{.}}/module-that-does-not-exists?ref=main",
+					ReasonPattern: "failed to vendor",
+				},
+			},
+		},
+		{
+			name: "module with 1 remote dependency",
+			layout: []string{
+				"g:module-test",
+				"g:another-module",
+			},
+			source: "git::file://{{.}}/module-test?ref=main",
+			configs: []hclconfig{
+				{
+					repo: "module-test",
+					path: "module-test/main.tf",
+					data: Module(
+						Labels("test"),
+						Str("source", "git::file://{{.}}/another-module?ref=main"),
+					),
+				},
+			},
+			wantFiles: map[vendorPathSpec]fmt.Stringer{
+				"git::file://{{.}}/module-test?ref=main#main.tf": Module(
+					Labels("test"),
+					Str("source", "{{index . 1}}"),
```

Comment:
hmm I usually like table tests, but in this case thinking if maybe we are stretching too much to do it, maybe N test functions (even with some duplication) would be easier to understand :thinking: (or a more sophisticated DSL, but then it may be way too much effort for this... IDK).

Reply:
In the beginning I was testing like this (several Test* funcs) but there's a lot of duplication even if you abstract in several functions.. for me it was worse, but I agree that the hacks here are not cool as well. But I have some ideas to improve this if we have time in the future.

Reply:
Hmm true, in the end a really good DSL wins in this case.